### PR TITLE
Delta XDS - improved

### PR DESCRIPTION
**Pull request by slonka**
This is a PR based on work from @sschepens https://github.com/envoyproxy/java-control-plane/pull/152 and @mikegajda https://github.com/HubSpot/java-control-plane/pull/4 .

At Allegro we've implemented it here: https://github.com/allegro/envoy-control/pull/255 .

We've seen 50-70% improvements in CPUs usage in Envoys that do not have significant traffic (mostly do XDS) and ~35% in more RPS intensive apps.

### CODE REVIEW

**Comment by slonka** [comment]
@mikegajda - some of your commits are missing sign-off, do you mind if I squash them and add you as a co-author?

**Comment by mikegajda** [comment]
Sure that'd be great, thanks @slonka!

File: cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java
Comment (slonka) [inline cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java:594]: @mikegajda is this used somewhere?

Comment (mikegajda) [reply cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java:594]: Hmmm, I don't see any usages either, so I think it's safe to remove

File: cache/src/main/java/io/envoyproxy/controlplane/cache/CacheStatusInfo.java
Comment (lukidzi) [inline cache/src/main/java/io/envoyproxy/controlplane/cache/CacheStatusInfo.java:20]: WDYT about making 2 different implementations `DeltaCacheStatusInfo` and `CacheStatusInfo` that implements same methods? We would have `lastWatchRequestTime` for both depends on implementation.

File: cache/src/test/java/io/envoyproxy/controlplane/cache/v2/SnapshotTest.java
Comment (lukidzi) [inline cache/src/test/java/io/envoyproxy/controlplane/cache/v2/SnapshotTest.java:130]: Should be fixex before merge

**Review by lukidzi — COMMENTED** [review]
Nice work! I think the only problem I have with changes is code duplication. We could provide more abstraction and work on implementation instead of adding `...Delta...`.

File: cache/src/test/java/io/envoyproxy/controlplane/cache/v2/SnapshotTest.java
Comment (mikegajda) [reply cache/src/test/java/io/envoyproxy/controlplane/cache/v2/SnapshotTest.java:130]: Sorry, this review slipped through my email. This should be an easy fix

Comment (mikegajda) [reply cache/src/test/java/io/envoyproxy/controlplane/cache/v2/SnapshotTest.java:130]: Fix is easy, but looks like I don't have commit privileges to this repo, and my fork of it is based on https://github.com/sschepens/java-control-plane/tree/delta-xds and hence very out of date. @slonka is it possible to get me privileges on this repo? Or this is a very small fix, should just be a matter of  substituting this for the commented out code:
```
     assertThat(snapshot.resources(CLUSTER_TYPE_URL))
        .containsEntry(CLUSTER_NAME, VersionedResource.create(CLUSTER))
        .hasSize(1);

    assertThat(snapshot.resources(ENDPOINT_TYPE_URL))
        .containsEntry(CLUSTER_NAME, VersionedResource.create(ENDPOINT))
        .hasSize(1);

    assertThat(snapshot.resources(LISTENER_TYPE_URL))
        .containsEntry(LISTENER_NAME, VersionedResource.create(LISTENER))
        .hasSize(1);

    assertThat(snapshot.resources(ROUTE_TYPE_URL))
        .containsEntry(ROUTE_NAME, VersionedResource.create(ROUTE))
        .hasSize(1);
```

**Comment by slonka** [comment]
FYI: I'm no longer an employee of Allegro and my involvement with Envoy related projects will decrease.

**Comment by mikegajda** [comment]
:wave: @lukidzi & @Ferdudas97 wanted to introduce myself here since @`slonka and I worked on this a bit before. I've picked up work on Envoy again and would love to see this get merged so that it can be used. What's the current status on this from your end, and is there anything I can do to help this progress?

**Comment by Ferdudas97** [comment]
> 👋 @lukidzi & @Ferdudas97 wanted to introduce myself here since @`slonka and I worked on this a bit before. I've picked up work on Envoy again and would love to see this get merged so that it can be used. What's the current status on this from your end, and is there anything I can do to help this progress?

Hi, work in this pr is probably done and we want to test it on our environment, but publishing snapshots is broken (I wrote about it on slack). `gpg: signing failed: unusable secret key`. Now, we are waiting for the secret key update.

**Review by kukulam-allegro — APPROVED** [review]

**Review by Automaat — DISMISSED** [review]

**Comment by Ferdudas97** [comment]
Tests on our environment are finished. We confirmed that, using delta improves cpu utilization ( dropped by ~50%).

**Comment by Ferdudas97** [comment]
Hi @rulex123, I tested it on our environment where we have ~1750 clusters. We are using [envoy-control,](https://github.com/allegro/envoy-control/pull/255) which is our abstraction layer over java-control-plane. According to our metrics, it returns 200-600 snapshots per minute.  
Zrzut ekranu 2022-03-16 o 07 54 47

On our service with enabled DELTA_GRPC and wildcard dependencies (service is receiving SoTW), we can see CPU utilization drop by 50%.

Zrzut ekranu 2022-03-16 o 08 24 36

Part of envoy's config 
```
dynamic_resources:
  lds_config:
    ads: {}
    resource_api_version: V3
    initial_fetch_timeout: 20s
  cds_config:
    ads: {}
    resource_api_version: V3
    initial_fetch_timeout: 20s
  ads_config:
    api_type: DELTA_GRPC
    transport_api_version: V3
    grpc_services:
      envoy_grpc:
        cluster_name: envoy-control-xds
```

**Comment by rulex123** [comment]
I would like to start removing support for xDS v2 altogether, do you think it would be possible to coordinate that effort with adding support for xDS incremental? i.e. If we removed support for v2 and then updated this patch accordingly, the amount of code to test/review to ship xDS incremental would reduce significantly.

**Comment by Automaat** [comment]
> I would like to start removing support for xDS v2 altogether, do you think it would be possible to coordinate that effort with adding support for xDS incremental? i.e. If we removed support for v2 and then updated this patch accordingly, the amount of code to test/review to ship xDS incremental would reduce significantly.

@rulex123 We don't want to merge these two topics, since these are two different topics. I have a local branch with almost removed v2 API and would like to push it as soon as we merge this pull request with delta XDS.

**Comment by rulex123** [comment]
> > I would like to start removing support for xDS v2 altogether, do you think it would be possible to coordinate that effort with adding support for xDS incremental? i.e. If we removed support for v2 and then updated this patch accordingly, the amount of code to test/review to ship xDS incremental would reduce significantly.
> 
> @rulex123 We don't want to merge these two topics, since these are two different topics. I have a local branch with almost removed v2 API and would like to push it as soon as we merge this pull request with delta XDS.

I agree we shouldn't merge the two in one patch: what I am suggesting is that we first ship removal of support for xDS v2, and after that change has been merged upstream, we rebase the `delta-xds-non-breaking-slonka-hash-bytes` branch on `main` so that the scope of the code to review/test in this patch is reduced.

I think this is worth it, especially if you have a branch where you started the v2 removal work. I can also help with that work, in case you don't have cycles to continue it. WDYT?

**Comment by Automaat** [comment]
> > > I would like to start removing support for xDS v2 altogether, do you think it would be possible to coordinate that effort with adding support for xDS incremental? i.e. If we removed support for v2 and then updated this patch accordingly, the amount of code to test/review to ship xDS incremental would reduce significantly.
> > 
> > 
> > @rulex123 We don't want to merge these two topics, since these are two different topics. I have a local branch with almost removed v2 API and would like to push it as soon as we merge this pull request with delta XDS.
> 
> I agree we shouldn't merge the two in one patch: what I am suggesting is that we first ship removal of support for xDS v2, and after that change has been merged upstream, we rebase the `delta-xds-non-breaking-slonka-hash-bytes` branch on `main` so that the scope of the code to review/test in this patch is reduced.
> 
> I think this is worth it, especially if you have a branch where you started the v2 removal work. I can also help with that work, in case you don't have cycles to continue it. WDYT?

I will work on removing API v2 during this week, hopefully I will prepare first PR this week

**Comment by mikegajda** [comment]
:wave: any updates here? I'd be happy to help get this merged as I'm hoping to deploy this change soon

**Comment by Ferdudas97** [comment]
> 👋 any updates here? I'd be happy to help get this merged as I'm hoping to deploy this change soon

Hey, the work is done on this branch. We are waiting for the review  and merging of this [PR](https://github.com/envoyproxy/java-control-plane/pull/221) -> there is a problem with publishing artifact for JCP.

**Comment by Ferdudas97** [comment]
@rulex123 @mikegajda it is ready for your reviews

File: server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java
Comment (mikegajda) [inline server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java:50]: Should this be `onV2StreamRequest` to fulfill this the goal above "throws if it sees a v2 request", or were you thinking of throwing on a V3 non-delta request here and the comment above is in out of sync?

Comment (Ferdudas97) [reply server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java:50]: it was out of sync

**Review by Automaat — APPROVED** [review]

**Pull request merged**
